README.md

```diff
@@ -1,634 +1,1855 @@
+market-driven interest-rate discovery.
+
+## Future Considerations
+
+### Alternative Credit History Method
+
+Currently, defaults are identified by whether or not a Borrower ID was burned in isolation of other
+active beacons. This has the unfortunate consequence of leaving a residual Finished Active UTxO
+after every full payment. Koios is about to rollout an
+[update](https://github.com/cardano-community/koios-artifacts/pull/269) that will enable another
+method for identifying defaults.
+
+By querying Koios' `tx_info` [endpoint](https://preprod.koios.rest/#post-/tx_info), the
+`plutus_contracts` field can be used to see what redeemer was used on each loan input. For example,
+when a Borrower ID is burned in the transaction, if the `MakePayment` redeemer is used to pay 7 ADA
+on a loan with a current outstanding balance of 7 ADA, this is a full payment. Meanwhile, if a
+Borrower ID is burned in the transaction and the `Unlock` or `SpendWithKeyNFT` redeemers are used,
+these are defaults. This method would enable borrowers to burn all remaining beacons during a full
+payment which would entirely eliminate the step dedicated to cleaning up Finished Active UTxOs.
+
+> There is a [bug](https://github.com/cardano-community/koios-artifacts/issues/275) in the current
+> version of Koios which is why this method cannot be used yet. The patch is set to be released with
+> the new version of Koios.
+
+Unfortunately, this method was discovered late into the development process. Due to time contraints
+with the Project Catalyst grants and the uncertainty around when the new version of Koios will be
+released, the decision was made to leave the current method in place. After the other protocols are
+updated, Cardano-Loans will be updated to use the new method. Koios should have released the new
+version by then.
+
+### DID Compatibility
+
+The current design may already be compatible with DIDs. If a government either allowed a user to
+register their Cardano staking credential or issued them a dedicated one, this protocol already
+natively supports DIDs. Lenders would just need to cross-reference the government's registered
+credentials when they query a borrower's credit history. However, if governments used NFTs instead
+of credentials as DIDs, the protocol may not already support DIDs.
+
+Therefore, whether any changes are needed for Cardano-Loans to support DIDs entirely depends on how
+DIDs are issued.
+
+### Term Extensions and Re-Negotiations
+
+Refinancing loans are a useful economic feature, and sometimes, borrowers just need a little more
+time to pay back a loan. These features were not implemented because they still require R&D. It is
+unclear whether these features need to be supported directly by Cardano-Loans or whether, thanks to
+its composability, a separate DeFi protocol will suffice. Using a separate protocol, that "plugs-in"
+to Cardano-Loans may result in less overall complexity than putting all of the features into one
+protocol.
```

Comment:
I don't think you need anything extra.  Why can't this just be a negotiation between the lender and borrower to pay off the previous loan and simultaneously open a new loan?  This would essentially roll the loan.  Or even a negotiation with a new lender to pay off the existing loan and generate a new one.  Cardano-loans can already do that.

Reply:
You currently cannot accept an offer in the same transaction where another loan is fully paid off because the beacon mint/burn count will not agree (partial payments are fine since there is no required mint/burn). The ideal way to make them agree is to have a single count, but this requires a single script to tally up the required beacon mint/burn count for all actions. Because of all of the logic required, it doesn't fit into a single script. Breaking it into multiple scripts introduces security vulnerabilities as well as some redundant executions. I have not found a method I am content with.

There are also questions that need to be addressed:
- How does a borrower signal their desire to refinance/re-negotiate? Dedicated beacons will likely be required.
- How does the lender know the borrower will use the new loan to pay off the old one? There is currently no guarantee for this.

Reply:
I don't understand the internal smart contract limitations well enough, but two things come to mind:
1. If there is a limit on beacon burning/creation caused by completely closing the loan, then maybe the lender could mandate paying off 99.99% such that it is then much more advantageous for the borrower to subsequently completely close out this last tiny bit to get the loan "off the books" and improve his credit score.
2. I was under the impression that when a lender signs a transaction that this can be composed of a chain of transactions such that they all atomically succeed or fail, enabled by inherent UTxO determinism.  Why can't the lender sign a transaction which amounts to a new loan as well as paying off (or 99.99%) the old loan?

Reply:
Paying off 99.99% could enable the composition, but currently, the borrower is the only person who can spend the loan UTxO in order to make a payment on it so it would ultimately have to be the borrower who submits the final refinancing transaction. It is also the borrower who accepts offers. The lender is not part of either of these transactions; they just create the Offer UTxOs.

With the way the contracts currently work, if a lender created a new offer with the intention of it being used to pay off 99.99% of a previous loan, how does the lender know the new principal will actually be used like this? Right now, there would be nothing stopping the borrower from accepting the second loan while still keeping the first. While this may increase the chances of the borrower defaulting, I don't think this is enough of an incentive for the borrower to actually use the new principal to pay off 99.99% of the old loan. New logic would be needed to enforce the proper usage of this new principal.

If lenders are fine with the risks, the protocol does currently support the borrower accepting the new offer and paying off 99.99% of the old loan in the same transaction. But there is currently no way for either the borrower or lender to broadcast their intention for refinancing. If the borrower created a new Ask UTxO for the same amount, it would just look like the borrower wants another loan. How would lenders tell the difference between a borrower who wants a second loan for 1000 ADA and another borrower who wants to refinance a loan for 1000 ADA? There is currently no way for lenders to distinguish between the two types of Asks. The same is true for Offers. Dedicated beacons would need to be added in order to distinguish between the two types of Asks/Offers.